intelmq/bots/experts/ripe/expert.py

```diff
@@ -15,177 +17,144 @@
         self.http_session.cert = self.ssl_client_cert
 
+    def __initialize_cache(self):
         cache_host = getattr(self.parameters, 'redis_cache_host')
         cache_port = getattr(self.parameters, 'redis_cache_port')
         cache_db = getattr(self.parameters, 'redis_cache_db')
         cache_ttl = getattr(self.parameters, 'redis_cache_ttl')
         if cache_host and cache_port and cache_db and cache_ttl:
-            self.cache = Cache(cache_host, cache_port, cache_db, cache_ttl,
-                               getattr(self.parameters, "redis_cache_password",
-                                       None)
-                               )
-
-    def query_ripe_stat_contact(self, resource):
-        cache_value = self.cache.get('stat:%s' % resource)
-        if cache_value and cache_value != CACHE_NO_VALUE:
-            return json.loads(cache_value)
-        elif cache_value == CACHE_NO_VALUE:
-            return []
-        response = self.http_session.get(self.URL_STAT_CONTACT.format(resource), data="",
-                                         timeout=self.http_timeout_sec)
-        if response.status_code != 200:
-            raise ValueError(STATUS_CODE_ERROR % response.status_code)
+            self.__cache = Cache(cache_host, cache_port, cache_db, cache_ttl,
+                                 getattr(self.parameters, "redis_cache_password", None))
 
+    def process(self):
+        with self.event_context() as event:
+            for target in {'source.', 'destination.'}:
+                abuse_key = target + "abuse_contact"
+                abuse = set(event.get(abuse_key).split(',')) if self.__mode == 'append' and abuse_key in event else set()
+
+                asn = event.get(target + "asn", None)
+                if asn:
+                    if self.__query['stat_asn']:
+                        abuse.update(self.__perform_cached_query('stat', asn))
+                    if self.__query['db_asn']:
+                        abuse.update(self.__perform_cached_query('db_asn', asn))
+
+                ip = event.get(target + "ip", None)
+                if ip:
+                    if self.__query['stat_ip']:
+                        abuse.update(self.__perform_cached_query('stat', ip))
+                    if self.__query['db_ip']:
+                        abuse.update(self.__perform_cached_query('db_ip', ip))
+                    if self.__query['stat_geo']:
+                        info = self.__perform_cached_query('stat_geolocation', ip)
+
+                        should_overwrite = self.__mode == 'replace'
+
+                        for local_key, ripe_key in self.GEOLOCATION_REPLY_TO_INTERNAL:
+                            if ripe_key in info:
+                                event.add(target + "geolocation." + local_key, info[ripe_key], overwrite=should_overwrite)
+
+                event.add(abuse_key, ','.join(abuse), overwrite=True)
+
+    @contextmanager
+    def event_context(self):
```

Comment:
This function should actually be moved into the `parent` class, where it should also be used.